Step3 - 볼링 점수판(점수 계산)

sehan opened the pull request:
볼링 점수계산 리뷰요청합니다. 

### 볼링 점수계산 ( FrameScoreCalculator )
볼링 점수계산 기능을 추가했습니다.
FrameScoreCalclulator 점수계산 역할을 가집니다.

잘 부탁드립니다. 😊

nio-d committed:
feat: 프래임 점수 계산 매소드 구현

> FrameScoreCalculator
>>> Frame 의 점수를 계산한다
>>> Frame 의 상태에 맞는 FrameScoreRule 을 적용하여 점수를 계산한다

nio-d committed:
refactor: BaseFrame 추가. NormalFrame, FinalFrame 중복제거

nio-d committed:
refactor: PinMarks 코드가독성 개선 , PinMark.blank 제거

nio-d committed:
bugfix: 9 프레임 Strike 시 점수 표기 오류수정

nio-d committed:
test: ScoreRule test 추가

nio-d committed:
refactor: 상수는 static final 로 변경

nio-d committed:
test: ScoreRule test 추가로 불필요해진 test 제거

nio-d committed:
test: 불필요한 단위테스트 제거

nio-d committed:
refactor: 사용하지 않는 매소드 제거

nio-d committed:
refactor: 코드정리

nio-d committed:
refactor: 매소드명 수정

javajigi reviewed (changes requested):
전체적인 객체설계, 클린코드 구현, 테스트 코드 구현 좋네요. 👍 
피드백 몇 개 남겼으니 확인해 보시고요.

강의 문서의 "볼링 점수판 리팩토링 힌트"와 제가 구현한 https://github.com/next-step/java-bowling/tree/javajigi 코드를 참고해 리팩토링할 부분이 있는지 검토해 볼 것을 추천합니다.

nio-d committed:
refactor: NormalPinMarker.mark() 매소드에서 분기처리 제거

nio-d committed:
refactor: FinalPinMarker.mark() 매소드에서 분기처리 제거

nio-d committed:
refactor: NormalPinMarker.toSigns() 에서 분기문 제거

nio-d committed:
refactor: FinalPinMarker.toSigns() 에서 분기문 제거

nio-d committed:
refactor: Frame 생성역할 분리

nio-d committed:
refactor: 클래스 이동, 네이밍 개선

nio-d committed:
refactor: Frame 점수계산시 조건에따른 분기처리 제거하고 상태에 따라 계산하도록 개선

FrameScoreRule.apply() 메소드를 제거하고 FrameScore 에 추가되어야할 이후 투구수 상태를 추가.
FrameScore 게산을 위해 추가해야할 횟수만큼 이후 Frame 에게 점수계산(이후 투구로 쓰러뜨린 핀수 점수에 반영) 을 요청하도록 처리

nio-d committed:
refactor: FrameScoreCalculatorFactory 삭제, FrameScoreFactory 로 역할을 대신

FrameScoreFactory
- NormalFrame, FinalFrame 별로 ScoreFactory 를 구현한다
- Frame 의 상태에 따라 FrameScore 객체를 생성하고 FrameScore 의 AdditionCounter 의 상태에 따라 next frame 의 핀수를 더 한다

nio-d committed:
test: AdditionCounterTest 추가

sehan commented on src/main/java/bowling/domain/FinalFramePinMarks.java:
```diff
@@ -19,72 +22,96 @@
     public List<PinMarkSign> toSigns() {
-        if (isSpare() && marks.size() == 2) {
+        if (isSpare() && getCountOfMarks() == 2) {
             return Arrays.asList(PinMarkSign.number(marks.get(0).getCountOfFallDownPins()), PinMarkSign.Spare);
```
marks.get() 이 부분은 정말 상당히 아주 많이 제거 하고 싶습니다. 😅
리팩토링 힌트를 보고 영감을 얻어 개선해 봤습니다.

sehan commented:
여러코드에 남아 있는 분기문들을 어떻게 리팩토링 해야하나 고민하다가 결국 "리팩토링힌트" 를 열어보고 영감을 얻었습니다 😊
Frame 과 PinMarker 그리고 FrameScore 에 사용된 분기문들을 제거해 봤습니다.

javajigi commented on src/main/java/bowling/domain/FinalPinMarker.java:
```diff
@@ -0,0 +1,260 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FinalPinMarker implements PinMarker {
```
NormalPinMarker에 있는 InProgress, Completed를 상속하는 inner 클래스들과 중복되는 부분이 많다.
PinMarker의 특성을 살펴보면 1 ~ 9 프레임과 10 프레임의 동작이 같다.
strike을 계산하려면 이후 2개의 값이 필요하고, spare는 1개, miss는 0개이다.
위와 같이 동작이 같다면 지금과 같이 분리하지 않고 같은 객체를 NormalFrame과 FinalFrame이 같은 방식으로 사용할 수 있지 않을까?

중복 제거를 위해 이 부분을 어떻게 개선할 수 있을 것인지 고민해 본다.
아무리 고민해봐도 해결책을 찾기 힘들다면 제가 구현한 https://github.com/next-step/java-bowling/tree/javajigi 코드를 참고해 힌트를 얻을 수도 있다.

javajigi reviewed (changes requested):
3단계에서 분기문을 제거하기 쉽지 않죠?
분기문을 제거하기 위해 노력한 흔적들이 많이 보이네요. 👍 
그럼에도 불구하고 PinMarker에 중복되는 코드가 많아 피드백 남겨봤어요.
지금 단계에서 한 단계 더 개선해 보면 어떨까요?

Olga-Latoken commented on src/main/java/bowling/domain/BaseFrame.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+abstract public class BaseFrame implements Frame {
```
Hi, javajigi, my name is Olga, I represent the international company LATOKEN, they are looking for Senior Java backend developer (remote) https://krasnoyarsk.hh.ru/vacancy/37741912, if you're interested, please, let me know, and hr-manager will contact you.
